### DISCUSSION TITLE
Problems implementing window close event in Unpackaged

### DISCUSSION THREAD

**Discussion by dimohy** [microsoft/WindowsAppSDK#2399] [Q&A]
WinUI 3 does not yet support window close event.

If you are developing an 'editing program' with WinUI 3, you should see a dialog asking if you want to save your edits before closing the window.

This should be implemented via the Closing event, but is currently not available in WinUI 3.

If I want to create a close event I need to intercept the window message and handle the close event to listen for the window close event.

Ok, so I implemented this using the win32 API. The package I used to do this is cswin32.

The following source code only intercepts window messages and does not process anything. So I think it should work normally, but it doesn't.

| Register WNDPROC to intercept
```csharp
        var wndProcFunc = new WNDPROC(WinProc);
        var prevWndFunc = PInvoke.SetWindowLongPtr((HWND)_hwnd, WINDOW_LONG_PTR_INDEX.GWL_WNDPROC, Marshal.GetFunctionPointerForDelegate(wndProcFunc));
        _prevWndFunc = Marshal.GetDelegateForFunctionPointer<WNDPROC>(prevWndFunc);
```

| WNDPROC func
```csharp
    private LRESULT WinProc(HWND hwnd, uint msg, WPARAM wParam, LPARAM lParam)
    {
        return PInvoke.CallWindowProc(_prevWndFunc, hwnd, msg, wParam, lParam);
    }
```

If you have experience implementing relevant features, please let me know what the problem is.

It doesn't seem to happen with Packaged.

But in Unpackaged, the program exits with an exception.
![image](https://user-images.githubusercontent.com/58620778/163522723-bf5eb3a3-e1d3-4b1b-b005-4d13dae33d9b.png)

ExecutionEngineException is thrown and it seems to be due to the way Windows App SDK handles Unpackaged, but I can't find the cause.

Below is the reproducible source code.

https://github.com/dimohy/csharp-check/tree/main/WinUI3/WinUI3UnpackagedWinProcTest

**Comment by castorix** [comment]
You can use **SetWindowSubclass** and **WM_CLOSE** :

![WinUI_WM_CLOSE](https://user-images.githubusercontent.com/22345506/163556823-fe67c43b-c4fc-4df2-b140-1395a2b610e2.gif)

```

private int WindowSubClass(IntPtr hWnd, uint uMsg, IntPtr wParam, IntPtr lParam, IntPtr uIdSubclass, uint dwRefData)
{
    switch (uMsg)
    {
        case WM_CLOSE:
            {
                Quit();
                return 0;
            }
            break;
    }
    return DefSubclassProc(hWnd, uMsg, wParam, lParam);
}

ContentDialogResult cdResult;
ContentDialog cd = null;
private async void Quit()
{
    if (cd == null)
    {
        cd = new ContentDialog
        {
            Title = "Question",
            Content = "Do you want to quit ?",
            PrimaryButtonText = "OK",
        };
        cd.PrimaryButtonText = "Yes";
        cd.SecondaryButtonText = "No";
        cd.XamlRoot = this.Content.XamlRoot;
        cdResult = await cd.ShowAsync();
        if (cdResult == ContentDialogResult.Primary)
            this.Close();
        else
            cd = null;
    }
}
```

Declarations :

```
        public delegate int SUBCLASSPROC(IntPtr hWnd, uint uMsg, IntPtr wParam, IntPtr lParam, IntPtr uIdSubclass, uint dwRefData);

        [DllImport("Comctl32.dll", SetLastError = true)]
        public static extern bool SetWindowSubclass(IntPtr hWnd, SUBCLASSPROC pfnSubclass, uint uIdSubclass, uint dwRefData);

        [DllImport("Comctl32.dll", SetLastError = true)]
        public static extern int DefSubclassProc(IntPtr hWnd, uint uMsg, IntPtr wParam, IntPtr lParam);

        public const int WM_CLOSE = 0x0010;


        IntPtr hWnd = IntPtr.Zero;
        private SUBCLASSPROC SubClassDelegate;
```

 In MainWindow constructor :

```
        hWnd = WinRT.Interop.WindowNative.GetWindowHandle(this);
        SubClassDelegate = new SUBCLASSPROC(WindowSubClass);
        bool bReturn = SetWindowSubclass(hWnd, SubClassDelegate, 0, 0);
```

**Reply by dimohy** [reply]
Thank you for your help.

Unfortunately, the same problem occurs.
What's more different is that even when packaged, an exception occurs.
I used Windows App SDK 1.0.2 and my environment is Windows 11.

**An exception occurs when resizing.**
![image](https://user-images.githubusercontent.com/58620778/163658495-e27dd4c0-45d0-4d69-919e-b649d379ae2f.png)

You can reproduce with a very simple sample below.

https://github.com/dimohy/csharp-check/tree/main/WinUI3/WinUI3UnpackagedWinProcTest

**Comment by castorix** [comment]
I cannot reproduce this with Windows App SDK 1.1.0-preview1 on Windows 10 21H1 (packaged or unpackaged)

**Reply by dimohy** [reply]
Can you test in Windows App SDK 1.0.2?

Among the various tests, I found a difference.

If you continue to resize the window, an exception occurs.

1. For Windows App SDK 1.1.0-preview1
- Executing the EXE file does not cause exceptions
- Exception when executed in debug mode (F5)

1. For Windows App SDK 1.0.2
- Executing the EXE file also creates an exception.
- Exception when executed in debug mode (F5)

**Reply by castorix** [reply]
> Can you test in Windows App SDK 1.0.2?

I cannot test with 1.0.2 (I spent time to clean the Runtimes after a problem and I only kept 1.1.0-preview1...)
But either with your project or mine (with all the code in **MainWindow.cs** instead of App.xaml.cs), impossible to make them crash, release or debug

**Comment by dimohy** [comment]
@castorix Thank you for your kind feedback.

> AppWindow.Closing event

I finally found a way. I found that AppWindow provides Closing events
(I don't remember it being there before. AppWindow didn't work properly on Windows 10, so I was out of the question for a while.)
AppWindow still has a null AppWindowTitlebar in Windows 10, but anyway, it's offering Closing events and it works fine!

The Closing event worked well and solved my current problem.

Anyway, in Windows App SDK 1.0.2 or 1.1.0-preview1,
SetWindowLongPtr and SetWindowSubclass did not succeed. This may be a bug in the Windows App SDK.

**Reply by jonwis** [reply]
Hey @rkarman - can you take a look?

@dimohy - please file an issue for us to track the potential bug.

**Reply by oliverw** [reply]
The hoops you have to jump through to get such basic stuff implemented in 2023 are mind-blowing. Seriously Microsoft what were you thinking.

**Reply by DarranRowe** [reply]
```
MainWindow::MainWindow()
{
	InitializeComponent();

	auto native_window = try_as<IWindowNative>();
	winrt::check_hresult(native_window->get_WindowHandle(&m_window_handle));
	m_window_id = winrt::Microsoft::UI::GetWindowIdFromWindow(m_window_handle);
	m_appwindow = winrt::Microsoft::UI::Windowing::AppWindow::GetFromWindowId(m_window_id);

	m_appwindow_closing_token = m_appwindow.Closing({ this, &MainWindow::AppWindow_Closing });
}
```
Devastating.
